### ISSUE TITLE
Support using jq for highly-expressive composition transformations

### ISSUE THREAD

**Issue by jutley** [crossplane/crossplane#4351]
### What problem are you facing?

I frequently find that the transformations offered by Crossplane are too crude for use cases that seem rather simple. The example I am running into now is that I would like to pass labels with a specific prefix onto the AWS resources created by a transformation. Using the prefix `example.com`, we may want to turn the following Kubernetes resource labels into the following AWS tags:

```yaml
# Kubernetes labels
example.com/foo: foo
example.com/bar: bar
something: else  # just to illustrate filtering
```

```yaml
# AWS tags
- key: example.com/foo
  value: foo
- key: example.com/bar
  value: bar
```

### How could Crossplane help solve your problem?

I would love to see Crossplane adopt some mechanism for highly-expressive transformations. I am particularly fond of jq, which has an [implementation written in golang](https://github.com/itchyny/gojq) – perfect for Crossplane.

For this use case, my patch may look like this:

```yaml
patches:
  - type: FromCompositeFieldPath
    fromFieldPath: metadata.labels
    toFieldPath: spec.forProvider.tags
    transforms:
      - type: jq
        jq:
          jqExpression: 'to_entries | map(select(.key | startswith("example.com/")))'
```

This is a rather simple expression and very effectively handles our use case.

We could alternatively build out functionality like this using a Composition Function, but I think this proposal provides a way that users could get a lot of power out of the box without needing to learn how Composition Functions work.

**Comment by turkenh** [comment]
This makes sense, however, I would rather prefer to revisit this after composition functions are promoted to beta with the updated design. I have a hunch that it will be a lot easier to achieve this with composition functions compared to the current state (i.e. alpha, running with xfn runner, etc).

**Comment by bobh66** [comment]
[json-e](https://json-e.js.org/) is another useful tool - I think there may be a generic case where a user may want to write their compositions with formatting from tool X (jq, json-e, JMESPath, yq, etc) embedded in it, and then run that through a composition function to process all of the references.  I discussed the possibility of running Composition Functions at various points in the composition process with @negz and it's something we might consider after the current iteration goes through.

**Comment by pedjak** [comment]
@jutley - [there is a similar idea](https://github.com/crossplane/crossplane/issues/2772#issuecomment-1589405740) - if we create a generic function - a wrapper around `yq`, we could do achieve the same functionalities, please have a look.

**Comment by jutley** [comment]
@pedjak That proposal could work, but I do want to make the argument to _not_ use a Crossplane Function.

In the example in that proposal, the user needs to be familiar with the structure of the Crossplane Function input. It's a leaky abstraction. In contrast, the transformation layer can operate on exactly the data selected, which I think has real usability advantages, even if it is less powerful.

Of course, there isn't any reason both couldn't be supported. The advantage of the Function is that someone could easily open-source it and anyone could use it without having native support.

Regarding `jq` vs. `yq`, my only concern is the feature gap between the two. I'm not sure where the more popular `yq` is these days (there is a `yq` written in python which actually wraps the `jq` binary, so full feature support), but the last time I tried it the gap was so large I found it to not be worth using, and I've heard similar complaints from my peers. If you are just looking for `yaml` compatibility, the `gojq` implementation of `jq` can handle YAML input and output.
